types/wlr_keyboard.c

```diff
@@ -109,14 +125,20 @@ void wlr_keyboard_notify_key(struct wlr_keyboard *keyboard,
 			event->state == WLR_KEY_PRESSED ? XKB_KEY_DOWN : XKB_KEY_UP);
 	}
 	keyboard_led_update(keyboard);
-	keyboard_modifier_update(keyboard);
+
+	bool updated = keyboard_modifier_update(keyboard);
+	if (updated) {
+		wl_signal_emit(&keyboard->events.modifiers, keyboard);
+	}
+
 	keyboard_key_update(keyboard, event);
 	wl_signal_emit(&keyboard->events.key, event);
 }
 
 void wlr_keyboard_init(struct wlr_keyboard *kb,
 		struct wlr_keyboard_impl *impl) {
 	kb->impl = impl;
+	kb->modifiers = calloc(1, sizeof(struct wlr_keyboard_modifiers));
```

Comment:
If there's a good enough reason to keep a pointer, this is missing a NULL check.

Reply:
I'll just get rid of the pointer.